Automattic-Tracks-iOS/Crash Logging/CrashLogging.swift

```diff
@@ -2,178 +2,251 @@ import Foundation
+    ///   - userInfo: A dictionary containing additional data about this error.
+    ///   - level: The level of severity to report in Sentry (`.error` by default)
+    func logError(_ error: Error, userInfo: [String: Any]? = nil, level: SentryLevel = .error) {
 
-     - Parameters:
-     - error: The error object
-     - userInfo: A dictionary containing additional data about this error.
-     - level: The level of severity to report in Sentry (`.error` by default)
-    */
-    static func logError(_ error: Error, userInfo: [String: Any]? = nil, level: SentrySeverity = .error) {
-        let event = Event(level: .error)
-        event.message = error.localizedDescription
+        let event = Event(level: level)
+
+        /// Use the unlocalized error message for better grouping
+        event.message = SentryMessage(formatted: (error as NSError).description)
+
+        /// If the developer provides their own userInfo, use that – otherwise read it from the Error
         event.extra = userInfo ?? (error as NSError).userInfo
+
+        /// Attach the localized description in case it has additional data
+        event.extra?["localized-description"] = error.localizedDescription
+
+        event.timestamp = Date()
+
+        SentrySDK.capture(event: event)
+        dataProvider.didLogErrorCallback?(event)
+    }
+
+    /// Writes a message to the Crash Logging system, and includes a stack trace.
+    ///
+    /// - Parameters:
+    ///   - message: The message
+    ///   - properties: A dictionary containing additional information about this error
+    ///   - level: The level of severity to report in Sentry (`.info` by default)
+    func logMessage(_ message: String, properties: [String: Any]? = nil, level: SentryLevel = .info) {
+
+        let event = Event(level: level)
+        event.message = SentryMessage(formatted: message)
+        event.extra = properties
         event.timestamp = Date()
 
-        Client.shared?.snapshotStacktrace {
-            Client.shared?.appendStacktrace(to: event)
+        SentrySDK.capture(event: event)
+        dataProvider.didLogMessageCallback?(event)
+    }
+
+    /// Sends an `Event` to Sentry and triggers a callback on completion
+    func logErrorImmediately(_ error: Error, userInfo: [String: Any]? = nil, level: SentryLevel = .error, callback: @escaping (Result<Bool, Error>) -> Void) throws {
+        try logErrorsImmediately([error], userInfo: userInfo, level: level, callback: callback)
+    }
+
+    func logErrorsImmediately(_ errors: [Error], userInfo: [String: Any]? = nil, level: SentryLevel = .error, callback: @escaping (Result<Bool, Error>) -> Void) throws {
+
+        var serializer = SentryEventSerializer(dsn: dataProvider.sentryDSN)
+
+        errors.forEach {
+            let event = Event(level: level)
+            event.message = SentryMessage(formatted: $0.localizedDescription)
+            event.timestamp = Date()
+            event.extra = userInfo ?? ($0 as NSError).userInfo
```

Comment:
```suggestion
        errors.forEach { error in
            let event = Event(level: level)
            event.message = SentryMessage(formatted: error.localizedDescription)
            event.timestamp = Date()
            event.extra = userInfo ?? (error as NSError).userInfo
```

Nitpick: I was surprised by the `$0` because I've always seen it in one liner closures. It might be just me...

Reply:
Works for me – changed in baec210c6d4af873da6fe2de90e6e15ebd727208